### ISSUE TITLE
RN 0.73+ compatibility issue: "Property 'addEventListener' does not exist on type 'RTCPeerConnection'"

### ISSUE THREAD

**Issue by IanMolldrem** [react-native-webrtc/react-native-webrtc#1544]
**Related to:**
- https://github.com/react-native-webrtc/react-native-webrtc/issues/1488
- https://github.com/react-native-webrtc/react-native-webrtc/issues/1466
- https://github.com/react-native-webrtc/react-native-webrtc/issues/1503

## Issue
```typescript
let pc: RTCPeerConnection = /*...*/;
pc.addEventListener("track", (event) => {});
```

```typescript
error TS2339: Property 'addEventListener' does not exist on type 'RTCPeerConnection'.
```

## Background
React Native's [TS config](https://github.com/facebook/react-native/blob/3098613f6e3514113b362e1a87d6cdf83cbbf520/packages/typescript-config/tsconfig.json) switched from a `node` (now known as [`node10`](https://www.typescriptlang.org/docs/handbook/modules/reference.html#node10-formerly-known-as-node)) to a [`bundler`](https://www.typescriptlang.org/docs/handbook/modules/reference.html#bundler) module resolution mode with `v0.73` ([old TS config found here](https://github.com/tsconfig/bases/blob/be6b3bb160889347b8614e8d18e1e88c40f8ecc9/bases/react-native.json)).

The `bundler` mode supports two new features which `node10` did not:
- [`package.json` "exports"](https://www.typescriptlang.org/docs/handbook/modules/reference.html#packagejson-exports)
- [`package.json` "imports"](https://www.typescriptlang.org/docs/handbook/modules/reference.html#packagejson-imports-and-self-name-imports)

`event-target-shim`, a dependency of this project, has an `"exports"` field in its [`package.json`](https://github.com/mysticatea/event-target-shim/blob/a37993c95ed9cc1d7cb89358ea126ed4f95372e8/package.json#L6):
```json
...
  "main": "index.js",
  "exports": {
    ".": {
      "import": "./index.mjs",
      "require": "./index.js"
    },
    "./es5": {
      "import": "./es5.mjs",
      "require": "./es5.js"
    },
    "./umd": "./umd.js",
    "./package.json": "./package.json"
  },
...
```

With RN's new TS config using module resolution mode `bundler`, this field is now taken into consideration. The [Node docs on entry points](https://nodejs.org/api/packages.html#packages_package_entry_points) describe how this field impacts entry resolution:
> The ["exports"](https://nodejs.org/api/packages.html#exports) provides a modern alternative to ["main"](https://nodejs.org/api/packages.html#main) allowing multiple entry points to be defined, conditional entry resolution support between environments, and preventing any other entry points besides those defined in ["exports"](https://nodejs.org/api/packages.html#exports). This encapsulation allows module authors to clearly define the public interface for their package.

So with resolution mode `node10`, which doesn't support the `"exports"` field, the entry point for a `event-target-shim` import is `index.js`, as defined by the `"main"` field. With mode `bundler`'s support of `"exports"`, and it taking precedence over `"main"`, entry point resolution is dynamically determined by the project's TS config's `"module"` field.

Resolution mode `bundler` requires `"module"` to be set to `es2015` or later (which it is, in RN's config). Where `commonjs` ([used in RN's previous TS config](https://github.com/tsconfig/bases/blob/be6b3bb160889347b8614e8d18e1e88c40f8ecc9/bases/react-native.json#L7)) results in `import { foo } from 'bar';` being transpiled to `const { foo } = require('bar');`, `es2015` has the opposite effect. 

## This is where I think the crux of the problem lies:
```typescript
import { defineCustomEventTarget } from 'event-target-shim';
```
### With module resolution mode `node`/`node10` and `"module"` of `commonjs`:
`event-target-shim` entry point resolved to `index.js`. `"types"` resolved to `index.d.ts`. Coincidentally, this project (`react-native-webrtc`) [uses this configuration](https://github.com/react-native-webrtc/react-native-webrtc/blob/master/tsconfig.json#L5).

### With module resolution mode `bundler` (`"module"` of `es2015` or later implied):
`event-target-shim` entry point resolved to `index.mjs`. There is no corresponding `index.d.mts`. If there were, and it had the same content as `index.d.ts`, the reported issue would not occur. Alternatively, `event-target-shim`'s `package.json` could be modified like so:
```json
...
  "main": "index.js",
  "exports": { 
    ".": {
-     "import": "./index.mjs",
+     "import": {
+       "default": "./index.mjs",
+       "types": "./index.d.ts"
+     },
      "require": "./index.js"
    },
...
```

Both of these "fixes" seem somewhat hacky to me, but they definitely work and may ultimately be best practice.

## RN `v0.73+` Workaround
Setting the [`resolvePackageJsonExports`](https://www.typescriptlang.org/tsconfig#resolvePackageJsonExports) compiler option to `false` in your  project's TS config will disable the [`package.json "exports"`](https://www.typescriptlang.org/docs/handbook/modules/reference.html#packagejson-exports) feature altogether.

---

@8BallBomBom @saghul, I'm not sure that your solution for this issue (added in https://github.com/react-native-webrtc/react-native-webrtc/pull/1515) is working as expected. If you bumped this project's TS version to `5.0.4` and modified the `tsconfig.json` to have the `bundler` resolution mode and `es2015` `"module"`, I suspect you'd see the same kind of issue described here.

My hunch is that if this project was distributed as transpiled `.js` files with no other changes (rather than `.ts` files like today), this issue would not occur. Due to your TS config, imports of `event-target-shim` such as:
```typescript
import { defineCustomEventTarget } from 'event-target-shim';
```

would be transpiled to:
```javascript
const { defineCustomEventTarget } = require('event-target-shim');
```
 
I assume this would then result in an `event-target-shim` entry point resolution of `index.js` (no TS errors) on RN `v0.73+`, with backwards compatability. There are probably other ways to fix this though.

---

## TL;DR
**RN `v0.73+` Workaround:**
Setting the [`resolvePackageJsonExports`](https://www.typescriptlang.org/tsconfig#resolvePackageJsonExports) compiler option to `false` in your  project's TS config should fix this issue 👍 

#### Platform Information  
* **React Native Version**:  0.73.6
* **WebRTC Module Version**: 111.0.3 + 118.0.4 + 118.0.5
* **Platform OS + Version**: Android API level 34, iOS 17.4

**Comment by 8BallBomBom** [comment]
After testing with a clean RN 0.73 project the issue appeared to be gone.
Ever since changing the import to be `event-target-shim/index` which you're not showing above ^
That being said i'll take another look 🤔

**Comment by IanMolldrem** [comment]
@8BallBomBom Thanks for taking a look! For what it's worth, this issue reproduces for me regardless of whether `event-target-shim/index` or `event-target-shim` is used for the import. 

Due to the TS config's module resolution mode `bundler` and `"module"`, `event-target-shim/index` still hits `index.mjs` as the entry point:
```json
...
".": {
      "import": "./index.mjs", // es2015
      "require": "./index.js" // commonjs
    },
...
```

Without the standard TS config shipped with RN `v0.73` (I'm on `0.73.6`), this may not occur.

**Comment by IanMolldrem** [comment]
I just reproed this by running `npx react-native init` (RN `0.73.6`), and installing this package 👍

**Comment by 8BallBomBom** [comment]
Alrighty 😓 
Will take a closer look in a little while.
Thanks for providing all the details and testing.

**Comment by joeporpeglia** [comment]
In my project this issue surfaces as a bundler error, possibly because I'm using the [metro patch recommended by the expo config plugin](https://github.com/expo/config-plugins/tree/main/packages/react-native-webrtc#event-target-shim). Bundling fails with the error `Error: Package subpath './index' is not defined by "exports"`. I'm working around this by stripping the `/index` path as shown below.

```
if (
    moduleName.startsWith('event-target-shim') &&
    context.originModulePath.includes('react-native-webrtc')
) {
    const updatedModuleName = moduleName.endsWith('/index')
        ? moduleName.replace('/index', '')
        : moduleName;

    const eventTargetShimPath = resolveFrom(context.originModulePath, updatedModuleName);

    return {
        filePath: eventTargetShimPath,
        type: 'sourceFile',
    };
}

return context.resolveRequest(context, moduleName, platform);
```

**Comment by 8BallBomBom** [comment]
Ahhhhhh yep, that is caused by the recent fix we put in place which appeared to work for some, myself included.
Changing the `metro.config.js` file may be the route to go as we likely don't want to change the tsconfig 🤔 
There is also the alternative, downgrading the package and having an absolute hell trying to have actual working types.
Nightmare on top of nightmare 😰

**Comment by chanphiromsok** [comment]
> In my project this issue surfaces as a bundler error, possibly because I'm using the [metro patch recommended by the expo config plugin](https://github.com/expo/config-plugins/tree/main/packages/react-native-webrtc#event-target-shim). Bundling fails with the error `Error: Package subpath './index' is not defined by "exports"`. I'm working around this by stripping the `/index` path in my metro config as shown below.
> 
> ```
> if (
>     moduleName.startsWith('event-target-shim') &&
>     context.originModulePath.includes('react-native-webrtc')
> ) {
>     const updatedModuleName = moduleName.endsWith('/index')
>         ? moduleName.replace('/index', '')
>         : moduleName;
> 
>     const eventTargetShimPath = resolveFrom(context.originModulePath, updatedModuleName);
> 
>     return {
>         filePath: eventTargetShimPath,
>         type: 'sourceFile',
>     };
> }
> 
> return context.resolveRequest(context, moduleName, platform);
> ```

this solution is work for me on expo 50.0.17 and livekit webrtc 114.1.1

**Comment by 8BallBomBom** [comment]
Will be looking to resolve this in a future update.

**Comment by IanMolldrem** [comment]
> Will be looking to resolve this in a future update.

Thanks @8BallBomBom, I appreciate it!

**Comment by nhonpassive** [comment]
Thank you @8BallBomBom , we appreciate it!

**Comment by prawee** [comment]
@joeporpeglia thank!! solved for me.

**Comment by 8BallBomBom** [comment]
Alrighty so it would seem the changes that we put in place do seem to actually work for some.
But then other conflicts are caused due to the expo metro patch.
Will keep this open for now as I'm still planning on looking deeper into the issue.

**Comment by saghul** [comment]
Closing since this is fixed now.

**Comment by dprevost-LMI** [comment]
Hey, I have just tried ` "resolvePackageJsonExports": false`, and it works, but that disabled it for all the modules in the project. Does someone have a better idea to have a fix that only affects `react-native-webrtc`?
Maybe the best would be to try to open a PR in react-native and update event-target-shim to v6 😅 ?

**Comment by IanMolldrem** [comment]
+1 @dprevost-LMI @saghul This isn't fixed for bare RN projects.

**Comment by saghul** [comment]
We are using it in a bare project with no issues.

Please share a small sample project where this fails.

**Comment by IanMolldrem** [comment]
@saghul I suspect you're not seeing the issue because this project's example project is still on RN 0.71.x. I'll get a sample project created.

**Comment by saghul** [comment]
Our main project is on 0.75.

**Comment by IanMolldrem** [comment]
@saghul Here's the sample project:
https://github.com/IanMolldrem/WebRTCRepro

Just created a new RN project and installed this package.

Repro is evident through the TS errors here:
https://github.com/IanMolldrem/WebRTCRepro/blob/main/App.tsx

The workaround detailed in my original issue description still works by uncommenting this:
https://github.com/IanMolldrem/WebRTCRepro/blob/f6cb28305dbf0534259c43254ff3b80fc0f88ba7/tsconfig.json#L4

**Comment by 8BallBomBom** [comment]
Might be worth revisiting #1468 🤔

**Comment by IanMolldrem** [comment]
@dprevost-LMI This patch of `event-target-shim@6.0.2` should resolve the issue in a more targeted way:
```patch
diff --git a/package.json b/package.json
index 045cf8bb2890a73e03d8aad727ab9f81bae2176a..d3639a901bb715dd5980482c84f818d8081dad05 100644
--- a/package.json
+++ b/package.json
@@ -5,7 +5,7 @@
   "main": "index.js",
   "exports": {
     ".": {
-      "import": "./index.mjs",
+      "import": "./index.js",
       "require": "./index.js"
     },
     "./es5": {

```

**Comment by dprevost-LMI** [comment]
❤ @IanMolldrem thanks for that. Since this issue is closed and you also provided a reproducible repo, it might be worth opening a new issue dedicated to all this. I can do it for you if you wish!

**Comment by 8BallBomBom** [comment]
Another issue can be opened but I'm not really sure if we would accept the posted patching method to be included 🤔
More likely to go down the road of #1468 or something similar.

**Comment by dprevost-LMI** [comment]
New issue open to fix the root cause instead of silencing with a tsc config: https://github.com/react-native-webrtc/react-native-webrtc/issues/1700#issue-3038071935
